src/v/kafka/server/handlers/fetch.cc

```diff
@@ -104,8 +108,24 @@ static ss::future<read_result> read_from_partition(
     std::unique_ptr<iobuf> data;
     std::vector<cluster::rm_stm::tx_range> aborted_transactions;
     try {
-        auto result = co_await rdr.reader.consume(
-          kafka_batch_serializer(), deadline ? *deadline : model::no_timeout);
+        auto failsafe_deadline = deadline ? *deadline + safeguard_fetch_timeout
+                                          : ss::lowres_clock::now()
+                                              + safeguard_fetch_timeout;
```

Comment:
here we might need to check if `deadline` is not set to `model::no_timeout`, if it is then the addition will cause overflow, this seems to happen in a test:

```
/vectorized/llvm/bin/../include/c++/v1/__chrono/duration.h:429:35: runtime error: signed integer overflow: 9223372036854775807 + 600000000000 cannot be represented in type 'rep' (aka 'long long')
```

Reply:
Arguably this method should take `model::timeout_clock::time_point` rather than `std::optional<model::timeout_clock::time_point>` as `deadline` since `model::no_timeout` is the standard way to represent "no timeout", and adding the optional just adds a 2nd way to do it which I don't think is necessary.

Reply:
Looks like this was partly done in https://github.com/redpanda-data/redpanda/pull/14924 replacement for this PR